### PR TITLE
micropython external module

### PR DESCRIPTION
**Pull request by bigbrett**
Adds a MicroPython external module enabling wolfSSL to be used as a back-end for MicroPython TLS and crypto operations. See the README.md for more details.

### CODE REVIEW

File: micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+#endif
+#include <wolfssl/wolfcrypt/settings.h>
+
+#include <wolfssl/wolfcrypt/aes.h>
+
+struct wolfssl_aes_ctx_with_key {
+    union {
+        Aes ctx;
+        struct {
+            uint8_t key[32];
```
Comment (JacobBarthelmeh) [inline micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c:72]: avoid magic numbers if possible AES_256_KEY_SIZE (32)

File: micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+            uint8_t keysize;
+        } init_data;
+    } u;
+    unsigned char iv[16];
```
Comment (JacobBarthelmeh) [inline micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c:76]: AES_IV_SIZE (16)

File: micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    }
+}
+
+STATIC void aes_final_set_key_impl(AES_CTX_IMPL *ctx, bool encrypt, mp_int_t block_mode) {
+    // first, copy key aside
+    uint8_t key[32];
+    uint8_t keysize = ctx->u.init_data.keysize;
+    memcpy(key, ctx->u.init_data.key, keysize);
+
+    assert(16 == keysize || 32 == keysize);
```
Comment (JacobBarthelmeh) [inline micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c:118]: AES_128_KEY_SIZE(16) AES_256_KEY_SIZE(32).

Any reason not to include AES_192_KEY_SIZE(24)?

File: micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    if (encrypt) {
+        wc_AesEncryptDirect(&ctx->u.ctx, out, in);
+    } else {
+        wc_AesDecryptDirect(&ctx->u.ctx, out, in);
+    }
+}
+
+STATIC void aes_process_cbc_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *out, size_t in_len, bool encrypt) {
+    if (encrypt) {
+        wc_AesCbcEncrypt(&ctx->u.ctx, out, in, in_len);
```
Comment (JacobBarthelmeh) [inline micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c:146]: Is there an error log we could print to in the case of these wolfSSL Aes Encrypt/Decrypt functions returning an error value?

File: micropython-modules/micropython-wolfssl/modussl_wolfssl.c [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+                mp_raise_ValueError(MP_ERROR_TEXT("invalid key"));
+                break;
+            default:
+                wolfssl_raise_error(err);
+                break;
+        }
+
+        wolfSSL_Cleanup();
+
+        // BRN-TODO Python error raising
```
Comment (JacobBarthelmeh) [inline micropython-modules/micropython-wolfssl/modussl_wolfssl.c:284]: Is this still needed here? the MBEDTLS macro/enum references

File: micropython-modules/micropython-wolfssl/ports/unix/user_settings.h [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+
+#undef  HAVE_TLS_EXTENSIONS
+#define HAVE_TLS_EXTENSIONS
+
+#undef  HAVE_SERVER_RENEGOTIATION_INFO
+#define HAVE_SERVER_RENEGOTIATION_INFO
+
+#undef  HAVE___UINT128_T
+#define HAVE___UINT128_T 1
```
Comment (JacobBarthelmeh) [inline micropython-modules/micropython-wolfssl/ports/unix/user_settings.h:183]: having UINT128 types is system dependent, not sure if all unix systems will have it

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]
Great work Brett! The tests ran for me:

```
$ ./run-tests.py -d /home/jak/Documents/osp/micropython-modules/micropython-wolfssl/tests/
pass  /home/jak/Documents/osp/micropython-modules/micropython-wolfssl/tests/wolfssl_keycert.py
pass  /home/jak/Documents/osp/micropython-modules/micropython-wolfssl/tests/wolfssl_poll.py
2 tests performed (3 individual testcases)
2 tests passed
```

Some minor feedback on the docs and magic numbers.

File: micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c (hunk 3 above)
Comment (bigbrett) [reply micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c]: was just mimicking the existing functionality without adding to it....shouldn't hurt to put it in though, so I'll add.

File: micropython-modules/micropython-wolfssl/README.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+## User settings
+wolfSSL can be customized for a specific target application or platform using a `user_settings.h` file. This module provides a default `user_settings.h` file for each port located in the `ports` directory. If you wish you further customise wolfSSL, you can create your own `user_settings.h` and pass it to the build system through the `WOLFSSL_USER_SETTINGS_FILE` variable. This will override the default one located in the `ports/$(WOLFSSL_PORT)` directory. The directory containing this file will be added to the compiler include path, so it is recommended you create a new directory holding this file with nothing else in it. Information on the various build options that can tailor wolfSSL to your application can be found in [chapter 2 of the wolfSSL manual](https://www.wolfssl.com/documentation/manuals/wolfssl/chapter02.html#features-defined-as-c-pre-processor-macro).
```
[the remark this answers is not in the document]
Comment (bigbrett) [reply micropython-modules/micropython-wolfssl/README.md]: fixed

File: micropython-modules/micropython-wolfssl/README.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+how to disable these modules for various ports. It is possible these instructions could become out of date if MicroPython makes changes to their build system. If the following steps don't work (if you get errors about multiple definitions for `uhashlib` and `ucryptolib` fucntions or types) then it is possible the build system has changed and you need to find another way to ensure the `MICROPY_PY_USSL`, `MICROPY_PY_UCRYPTOLIB` and `MICROPY_PY_UHASHLIB` C macros are all defined to zero at build time. `grep` will be your friend here...
+
+### unix
+For the unix port, the three C macros above can be easily disabled by setting the `MICROPY_PY_USSL=0` makefile variable in `ports/unix/mpconfigport.mk`. This should automatically ensure the `MICROPY_PY_USSL` C macro is not set, as well as prevent the `MICROPY_PY_UCRYPTOLIB` and `MICROPY_PY_UHASHLIB` C macros  from later being defined to 1 in `port/unix/variants/mpconfigvariant_common.h`. `MICROPY_TRACKED_ALLOC` should be set to 1 by default and no further action is needed.
```
[the remark this answers is not in the document]
Comment (bigbrett) [reply micropython-modules/micropython-wolfssl/README.md]: fixed

File: micropython-modules/micropython-wolfssl/ports/unix/user_settings.h (hunk 6 above)
Comment (bigbrett) [reply micropython-modules/micropython-wolfssl/ports/unix/user_settings.h]: good catch, fixed

File: micropython-modules/micropython-wolfssl/README.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
+# Overview
+ `micropython-wolfssl` is a collection of [user modules](https://docs.micropython.org/en/v1.19.1/develop/cmodules.html) for [micropython](https://micropython.org/) that enable [wolfSSL](https://www.wolfssl.com) to be used for TLS and cryptographic operations. It is an API-compatible replacement for the built-in implementations provided by MicroPython's `ussl`, `ucryptolib` and `uhashlib` modules.
+
+# How it works
+`micropython-wolfssl` provides three modules: `wolfssl`, `wolfcryptolib`, and `uhashlib`. These modules implement the same API as MicroPython's built-in implementations of of SSL/TLS (`ussl`), encryption (`ucryptolib`), and hashing (`uhashlib`). This means that python code utilizing these modules should not need to be changed (other than the name of the imported modules) when `micropython-wolfssl` is used, and all MicroPython documentation for these modules can be followed. The only changes in behavior are different SSL error codes returned on failure. WolfSSL has much more comprehensive error codes than most other SSL libraries, and are detailed in [appendix F of the wolfSSL manual](https://www.wolfssl.com/documentation/manuals/wolfssl/appendix06.html).
```
[the remark this answers is not in the document]
Comment (bigbrett) [reply micropython-modules/micropython-wolfssl/README.md]: good catch, fixed

File: micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c (hunk 1 above)
Comment (bigbrett) [reply micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c]: fixed

File: micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c (hunk 2 above)
Comment (bigbrett) [reply micropython-modules/micropython-wolfssl/moducryptolib_wolfssl.c]: fixed

**Comment by bigbrett** [comment]
@JacobBarthelmeh Thanks a ton for the feedback. Believe I addressed everything. Retested with changes on both unix and STM32

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**